File: snippets/python/pick-allowed-update-fields-from-a-patch-body-with-python/src/pick_allowed_patch_fields.py
from typing import Dict, Iterable, Mapping, Set


def pick_allowed_patch_fields(
    patch_body: Mapping[str, object], allowed_fields: Iterable[str]
) -> Dict[str, object]:
    """Return only allowed top-level fields from a PATCH-style mapping."""
    # Flow:
    #   read the incoming patch body and allowed field list
    #      |
    #      +-> matching top-level keys -> keep them with original values
    #      `-> unknown keys -> ignore them in the returned patch
    if not isinstance(patch_body, Mapping):
        raise TypeError("patch_body must be a mapping")

    allowed_field_names = _normalize_allowed_fields(allowed_fields)
    return {
        key: value
        for key, value in patch_body.items()
        if isinstance(key, str) and key in allowed_field_names
    }


def _normalize_allowed_fields(allowed_fields: Iterable[str]) -> Set[str]:
    normalized = set()
    for field_name in allowed_fields:
        if not isinstance(field_name, str) or field_name.strip() == "":
            raise ValueError("allowed field names must be non-empty strings")
        normalized.add(field_name.strip())
    return normalized


if __name__ == "__main__":
    filtered_patch = pick_allowed_patch_fields(
        {
            "display_name": "Ada",
            "bio": None,
            "role": "admin",
        },
        ["display_name", "bio"],
    )
    print(filtered_patch)

File: snippets/go/pick-allowed-fields-from-patch-body-with-go/pick_allowed_patch_fields.go
package main

import (
	"errors"
	"log"
	"strings"
)

// Flow: inspect the PATCH body -> keep only allowlisted top-level keys -> preserve explicit values and ignore unknown fields.
func PickAllowedFieldsFromPatchBody(body interface{}, allowedFields []string) (map[string]interface{}, error) {
	allowed := make(map[string]struct{}, len(allowedFields))
	for _, field := range allowedFields {
		trimmed := strings.TrimSpace(field)
		if trimmed == "" {
			return nil, errors.New("allowed fields must not be blank")
		}
		allowed[trimmed] = struct{}{}
	}

	if body == nil {
		return map[string]interface{}{}, nil
	}

	patchBody, ok := body.(map[string]interface{})
	if !ok {
		return nil, errors.New("patch body must be a map[string]interface{}")
	}

	picked := make(map[string]interface{})
	for key, value := range patchBody {
		if _, ok := allowed[key]; ok {
			picked[key] = value
		}
	}

	return picked, nil
}

func main() {
	body := map[string]interface{}{
		"display_name": "Ada",
		"role":         "admin",
		"nickname":     nil,
	}

	picked, err := PickAllowedFieldsFromPatchBody(body, []string{"display_name", "nickname"})
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("picked=%v", picked)
}
